### ISSUE TITLE
Configurable associated product image not showing

### ISSUE THREAD

**Issue by rani-priya** [magento/magento2#24571]
### Preconditions (*)

1. Magento 2.3-develop

### Steps to reproduce (*)

1. create a configurable product based on _**Text Swatch attribute**_
2. check it at frontend

### Expected result (*)

1. product image should be change as per selected configuration

### Actual result (*)

1. product image not showing as per selected configuration

**Example**:
https://github.com/magento/magento2/issues/24571#issuecomment-537949454

**Comment by engcom-Bravo** [comment]
Hello @rani-webkul. Probably we miss or misunderstand something, so please watch our gif and tell us where we go wrong. Thank You.
![conf](https://user-images.githubusercontent.com/51680745/64793134-3e0d4580-d583-11e9-8208-143faf91b9f0.gif)

**Comment by rani-priya** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @rani-webkul. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @rani-webkul, here is your [Magento instance](https://i-24571-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-24571-2-3-develop.instances.magento-community.engineering/admin](https://i-24571-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by rani-priya** [comment]
@engcom-Bravo please look into https://i-24571-2-3-develop.instances.magento-community.engineering/test-conf.html

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-20134` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by ihor-sviziev** [comment]
Hi @rani-webkul or @engcom-Bravo ,
Could you add some gif/video that shows how issue reproducing? Unfortunately test instance already not available to check how @rani-webkul configured it.

**Comment by rani-priya** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @rani-webkul. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @rani-webkul, here is your [Magento instance](https://i-24571-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-24571-2-3-develop.instances.magento-community.engineering/admin](https://i-24571-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by rani-priya** [comment]
@ihor-sviziev, Please check https://i-24571-2-3-develop.instances.magento-community.engineering/test-config.html

**Comment by ihor-sviziev** [comment]
Hi @rani-webkul,
I just checked - param "Update Product Preview Image" is set to "No" in Stores >> Attributes >> Product >> size
![image](https://user-images.githubusercontent.com/1873745/64841016-2aef8980-d606-11e9-9a3a-70c9ece3c5e2.png)


That's why it didn't appeared.
Once I changed it to "Yes" and flushed cache - product image started changing

**Comment by rani-priya** [comment]
@ihor-sviziev I have assigned images to each associated product but it's not showing at frontend why?

**Comment by ihor-sviziev** [comment]
Hi @rani-webkul 

> @ihor-sviziev I have assigned images to each associated product but it's not showing at frontend why?
I just checked - image URL generated correctly, but somehow resized image shows default magneto thumbnail:

https://i-24571-2-3-develop.instances.magento-community.engineering/pub/media/catalog/product/cache/f77037f53323565848a245b9388a2a70/i/m/img-3_1.png
![image](https://user-images.githubusercontent.com/1873745/64841622-4d82a200-d608-11e9-84b4-a3987f7d7658.png)

While without "cache" it shows correct image:
https://i-24571-2-3-develop.instances.magento-community.engineering/pub/media/catalog/product/i/m/img-3_1.png
![image](https://user-images.githubusercontent.com/1873745/64841644-5d01eb00-d608-11e9-95b7-f02b5b6b066a.png)

I'm not sure why preview image wasn't generated correctly - it might be issue with test instance, or another issue, but image URL is correct, so swatches functionality works fine.

Could you confirm that everything works fine, or add some more details?

**Comment by engcom-Bravo** [comment]
Hi, @ihor-sviziev. Thank You for Your contributing to this issue. With the aforementioned setting set to Yes everything works fine. Maybe we really ought to close this issue. But one thing still bothers me - is it just an inaccuracy in the settings comment or not: "Filtering by this attribute will update the product image on catalog page". Usually filtering is associated with the layered navigation. And even if we do not consider that there is one more thing - the catalog page is usually stands for category page where alongside the current product there are other products presented as well. In this issue we've been testing product's behavior on the product page (not category), choosing (not filtering, though one may say in essence filtering and choosing are the same things) among available configurable variations. Unfortunately I didn't find anything on this setting in the official documentation but I found this on some other recourse: ["...tick the Update Product Preview Image, so your Magento 2 visual swatches could be displayed on the category page](https://www.mag-manager.com/product-information/magento-product-management/add-magento-2-visual-swatches-with-store-manager-for-magento/). And it sustains my opinion on our setting to be designated for changing images when filtering through the layered navigation on a category page. And as to the product page it looks quite natural that the images here should be swapped by default when one clicks through different swatches. Just like in case with the [attribute type drop-down ](https://github.com/magento/magento2/issues/24571#issuecomment-530855159) where this setting is even unavailable at all.
![attrs](https://user-images.githubusercontent.com/51680745/64849353-e53dbb80-d61b-11e9-9520-1f14bc4f7415.gif)
So what do You think about that?

**Comment by ihor-sviziev** [comment]
It really looks confusing, it’s better to change the comment. Anyway it
looks separate issue and another actual & expected results.
Also need to check if this option affects any other places (according to
name it shouldn’t, but according to comment - it might.

**Comment by engcom-Bravo** [comment]
@ihor-sviziev this issue is actual and correct. There's a simple logical prove to it. As I've already mentioned the Dropdown attribute, for example, doesn't even have "Update Product Preview Image" setting and it updates images by default. I even haven't tested what I'm going to talk about now because it's enough to just imagine it in order to understand that any further logic will fail with it. So lets assume a situation You have two attributes a dropdown one and a text swatch one. The last one has "Update Product Preview Image" set to "No". Now You have a configurable product created based on those two attributes with the separate images assigned to its variations. Go to the configurable product's page and try chose a concrete variation. Shall that variation's image appear or not? From one point of view it shall as there is a dropdown attribute involved, from the other point of view though it shan't as there is a text swatch attribute with "Update Product Preview Image" set to "No" involved. One logic excludes the other one while they are supposed to act simultaneously. Therefore no matter how it works in reality it works wrong (of course if there do not appear some warning messages on that problem when creating a configurable product. But I don't think there do any). So I'm reconfirming the issue. If You see my reasoning wrong I'm open for discussion.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-20134` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by ihor-sviziev** [comment]
Sure, let’s improve it!
Just wanted to link following PR that introduced this issue https://github.com/magento/magento2/pull/19184

**Comment by engcom-Bravo** [comment]
Beautiful - here we have a prove that "Update Product Preview Image" originally didn't affect the Product page but the Category page as due to User Guide and the comment "Filtering by this attribute will update the product image on catalog page" was quite correct. So as I see everything had been working just as intended before #16446 appeared, was confirmed and then "fixed" by #19184. What do You think about it?

**Comment by ihor-sviziev** [comment]
Hi @akaplya,

I think we need your point of view as you're assigned to Swatches component.

Could you review this issue and add your opinion?

Is it ok that "Update Product Preview Image" now affects Product page once #19184 was merged or should we revert it? If we should revert it - maybe it's good idea to adjust option name or comment below it?

@engcom-Bravo I think before this feedback we can't say that this is "confirmed" issue and we can start working on it.

**Comment by engcom-Bravo** [comment]
You are doing extremely correct @ihor-sviziev) And that's good!

**Comment by rani-priya** [comment]
@ihor-sviziev can we work on this issue?

**Comment by ihor-sviziev** [comment]
> @ihor-sviziev can we work on this issue?

Hi @rani-webkul,
Currently it’s not clear what is expected result, @akaplya should define it

**Comment by rani-priya** [comment]
@akaplya we need your point of view, please check this issue and confirm us what to do

**Comment by rani-priya** [comment]
Hi @ihor-sviziev,
we are not getting any update from @akaplya, I think we should discuss it with other member.

**Comment by ihor-sviziev** [comment]
Hi @rani-webkul,
I'll try to push it forward. 

I created issue in architecture repository, hope it will speed up response https://github.com/magento/architecture/issues/309

**Comment by engcom-Bravo** [comment]
Hi @ihor-sviziev. I've been seeing Your efforts to push the deal forward but time goes on and we're still there with wrong behavior persisting in Magento. With the considerably wrong behavior. So let me demonstrate it with the real example. 
First, let me appeal to the bare formal logic once more as I already did it in the [comment](https://github.com/magento/magento2/issues/24571#issuecomment-532101204) - the images for the configurable variations can not be `updated` and `not updated` at the same time if one of the attributes is set to update the image while the other - to not update it. And of course they are not - image is updated if You pick-up the updating attribute and is not when You pick-up the non-updating one. In half of the cases (the updating attribute is chosen first) it will produce a serious inadequacy. 
So there's a configurable product "Background Colored Letter" with four variations based on Text Swatch and Dropdown attributes. Here the variations go:
A-Red  - 101$;
A-Blue - 102$;
B-Red  - 103$;
B-Blue - 104$.
Something like this:
![confandvar](https://user-images.githubusercontent.com/51680745/66125872-b8643f00-e5f0-11e9-9136-d514ef92d7c9.png)
Here we enter the configurable's Product Page with no attributes selected:
![config](https://user-images.githubusercontent.com/51680745/66125966-f2354580-e5f0-11e9-9c0f-a07efca0909a.png)
Now assume a customer choses "Red" from the dropdown menu and  "B" from the swatch. Evidently we expect the customer will get B-Red's image with price 103$, i.e. this (watch attributes selected):
![bredright](https://user-images.githubusercontent.com/51680745/66126612-87850980-e5f2-11e9-8704-479bf8a8212d.png)
But instead he or she gets this (watch the same attributes selected):
![bredwrong](https://user-images.githubusercontent.com/51680745/66126694-bc915c00-e5f2-11e9-9d1b-51b4f4c03667.png)
Notice by the way the price is reflected correctly.
If at least from some point of view that can be deemed an expected behavior, (kindly ask You to excuse me if I letting my self here something more than is generally allowed), then further we can call any behavior an expected one, regardless of how it is consistent with elementary logic.
The current situation may lead to a customer has bought a product not the one that he or she were thinking he or she were buying. And that's definitely not good.
Of course image updating performance on Product Page could be arranged in other logically correct ways like, for example, it can be taken out from the attribute and put among the settings of Configurable Product. So the image is ether updated due to selected variation or is not updated at all remaining the Configurable's one. But it never can be `updated` and `not updated` simultaneously. 
I will hope this post with a visual demonstration of the problem will foster the revert of the wrong commit. With that the issue remains in fact Confirmed but I accept the relative label is better not be on it as the contributors will strive to fix it with additional fix, while it definitely needs just a revert of the already existing one.

**Comment by ihor-sviziev** [comment]
Hi @engcom-Bravo,
Thank you for your input! With your example it’s becoming really clear that we have to do following:
1. Revert changes that were introduced in https://github.com/magento/magento2/pull/19184
2. Describe in https://github.com/magento/magento2/issues/16446 that that’s not an issue and related PR changes will be reverted
3. Update description for this option to reflect that it’s used only for category/search pages
4. We need to add MFTF test that will prevent appearing of this issue in future

Are you agree with me?

**Comment by engcom-Bravo** [comment]
@ihor-sviziev, I totally agree with You.

**Comment by engcom-Bravo** [comment]
Of course @ihor-sviziev we must understand that after #19184 had been merged there as well could be merged other commits that are dependent or rely on the behavior created by #19184. And there is some peril that those so to say bounded commits may throw then some new unexpected behavior. So if You could do some additional check in that relation before those fore steps are undertaken it would be great.

**Comment by pemann** [comment]
@magento give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @pemann. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @pemann, here is your [Magento instance](https://i-24571-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-24571-2-3-develop.instances.magento-community.engineering/admin](https://i-24571-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Delta 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-20134` were created

**Issue Available**: @engcom-Delta, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by ukguy** [comment]
We have a similar issue with attributes. All the images appear in the gallery but when selecting for eg, yellow option from a dropdown, it just selects a magento logo, not the yellow image.
When inspecting the filename of the yellow image it has _ underscores.
When selecting the filename of the magento logo the url is correct but it has %20 instead of underscores.

**Comment by ukguy** [comment]
Based on my previous comment it seems it's not selecting the image due to a bug with the underscores, so it doesn't select the image which is there. Maybe a javascript error?

**Comment by ihor-sviziev** [comment]
@ukguy it looks like different issue, please report it separately

**Comment by ukguy** [comment]
Ok 👍

**Comment by sandy6666** [comment]
@rani-webkul are you using any 3rd party plugin

**Comment by chinmayphutela** [comment]
@rani-webkul  or @ihor-sviziev 
Were you guys able to solve this configurable product - swatch image issue ?

I am facing the same issue and would love to understand how to rectify it. I am not a developer so would love if you could explain it in simple terms.

**Comment by ajithkumar-maragathavel** [comment]
@magento give me 2.3.4 instance

**Comment by magento-engcom-team** [comment]
Hi @ajithkumar-maragathavel. Thank you for your request. I'm working on Magento 2.3.4 instance for you

**Comment by magento-engcom-team** [comment]
Hi @ajithkumar-maragathavel, here is your [Magento instance](https://i-24571-2-3-4.instances.magento-community.engineering).
Admin access: [https://i-24571-2-3-4.instances.magento-community.engineering/admin_4e69](https://i-24571-2-3-4.instances.magento-community.engineering/admin_4e69)
Login: `d85cf5c1` Password: `5dff328627f7`
Instance will be terminated in up to 3 hours.

**Comment by sarbjeet-cmsDev** [comment]
I had same issue(on category page, on click swatch images were not changing )
I found my custom theme etc/view.xml was missing
<image id="product_swatch_image_large" type="image">
                <width>320</width>
                <height>270</height>
            </image>
            <image id="product_swatch_image_medium" type="thumbnail">
                <width>320</width>
                <height>270</height>
            </image>
            <image id="product_swatch_image_small" type="small_image">
                <width>320</width>
                <height>270</height>
            </image>

**Comment by ihor-sviziev** [comment]
@sidolov @sivaschenko could you explain why this issue was closed without any comments?

**Comment by sivaschenko** [comment]
The issue cannot be reproduced. Attribute option "Update Product Preview Image" must be set to "Yes" if you want to change the image for a configurable product with text swatch options. 
![image](https://user-images.githubusercontent.com/2028541/98972289-24121e80-250a-11eb-941f-8006135c9234.png)

**Comment by eBusinessAtlantique** [comment]
> <image id="product_swatch_image_large" type="image">
>                 <width>320</width>
>                 <height>270</height>
>             </image>
>             <image id="product_swatch_image_medium" type="thumbnail">
>                 <width>320</width>
>                 <height>270</height>
>             </image>
>             <image id="product_swatch_image_small" type="small_image">
>                 <width>320</width>
>                 <height>270</height>
>             </image>

Thanks for that reply, this was the solution.
But the appearance of the swatch image is quite slow. Why ?
Thanks

**Comment by Fidelity88** [comment]
> I had same issue(on category page, on click swatch product thumbnail was not changing )
> I found my custom theme etc/view.xml was missing
> 
> ```
> <image id="product_swatch_image_large" type="image">
>                 <width>320</width>
>                 <height>270</height>
>             </image>
>             <image id="product_swatch_image_medium" type="thumbnail">
>                 <width>320</width>
>                 <height>270</height>
>             </image>
>             <image id="product_swatch_image_small" type="small_image">
>                 <width>320</width>
>                 <height>270</height>
>             </image>
> ```

Almost a year later and I can say, this also was the fix for my theme! Thanks for sharing this!!

**Comment by junaidjavaid** [comment]
In my case hide product from product page is working for simple products but not working for configureable product. i have selected hide option in backend for some images but of configureable product but noyt working on frontend.
